### PR TITLE
Disk cache

### PR DESCRIPTION
**Pull request by abergeron**
This add an on-disk cache for kernels which speeds up recompilation.

From my light testing it shaves off about 1 second of a 3 seconds program that compiles an elemwise kernel and runs it, which means that it should be a net gain in most situations.

The cache is disabled by default and enabled when the environment variable GPUARRAY_CACHE_PATH is set to some path.  The cache should be fine for use over NFS with multiple processes.  Entries that were writing while a system crash happens can get corrupted or lost, but the cache should recover by ignoring and replacing bad entries. There is currently no built-in pruning mechanism, but there is a python script which can act as such.  It is safe to run it while other processes are accessing the cache.

If there is a problem accessing the given cache path, the cache will be disabled.

This still needs testing on Windows and macOS to make sure it works.  Especially on windows the semantics for touching files that are opened by other processes is significantly different, which may require adjustment.

Finally DO NOT MERGE until the 0.6 release is out the door.  This can be part of an eventual 0.6.1 release when it has been properly tested.

### CODE REVIEW

File: src/util/skein.c [hunk 1]
```diff
@@ -0,0 +1,309 @@
+/***********************************************************************
```
Comment (nouiz) [inline src/util/skein.c]: I didn't review this commit.

File: src/gpuarray/buffer.h [hunk 2]
```diff
@@ -328,9 +328,9 @@ GPUARRAY_PUBLIC int gpudata_move(gpudata *dst, size_t dstoff,
  * \returns the new buffer in dst_ctx or NULL if no efficient way to
  *          transfer could be found.
  */
-GPUARRAY_LOCAL int gpudata_transfer(gpudata *dst, size_t dstoff,
-                                    gpudata *src, size_t srcoff,
-                                    size_t sz);
+GPUARRAY_PUBLIC int gpudata_transfer(gpudata *dst, size_t dstoff,
```
Comment (nouiz) [inline src/gpuarray/buffer.h]: What is the consequense of this? SHould we bump version number? How this affect Theano?

File: src/gpuarray/config.h [hunk 3]
```diff
@@ -12,19 +12,15 @@
   #else
    #define GPUARRAY_PUBLIC __declspec(dllimport)
   #endif
-  #define GPUARRAY_LOCAL
```
Comment (nouiz) [inline src/gpuarray/config.h]: Should this commit and the previous should be included in 0.6? If so, make a new PR for this.

File: bin/gpuarray-cache [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    else:
+        mult = 0
+    return int(num * mult)
+
+
+if __name__ == '__main__':
+    import argparse
+
+    parser = argparse.ArgumentParser(description='libgpuarray cache maintenance utility')
+    parser.add_argument('-s', '--max_size', help='Set the maximum size for pruning')
```
Comment (nouiz) [inline bin/gpuarray-cache]: The help isn't clear. Does it mean we keep that number of element? Does this take into account the date? creation/usage date?

File: bin/gpuarray-cache [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+
+import os
+
+def clean(max_size):
+    content = []
+    for root, dirs, files in os.walk(os.environ.get('GPUARRAY_CACHE',
+                                                    '~/.gpuarray/cache/')):
```
Comment (nouiz) [inline bin/gpuarray-cache]: Why set a default here when if GPUARRAY_CACHE isn't set, we don't have a default?

If it work well on Linux with NFS, I would enable it by default on Linux. Until it is more tested on mac and windows, we should wait for those platform.

File: bin/gpuarray-cache (hunk 4 above)
Comment (nouiz) [inline bin/gpuarray-cache]: I understood the size as number of element. You probably mean by disk space. Be more clear and give the unit.

File: bin/gpuarray-cache [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    content.sort()
+    cur_size = 0
+    for _, size, path in content:
+        cur_size += size
+        if cur_size > max_size:
+            os.remove(path)
+
+
+SUFFIXES = {'B': 1, 'K': 1 << 10, 'M': 1 < 20, 'G': 1 << 30, 'T': 1 << 40,
```
Comment (nouiz) [inline bin/gpuarray-cache]: M is missing a <

File: bin/gpuarray-cache [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            'P': 1 << 50, 'E': 1 << 60, 'Z': 1 << 70, 'Y': 1 << 80}
```
Comment (nouiz) [inline bin/gpuarray-cache]: I would not note this scrip to survive up to Y size for the cache alone! Talk about being future proof!

File: bin/gpuarray-cache [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def get_size(s):
+    i = 0
+    while i < len(s) and (s[i].isdigit() or s[i] == '.'):
+        i += 1
```
Comment (nouiz) [inline bin/gpuarray-cache]: This is too complicated then needed, but it work. Just doing if s[-1] in "BKM"... would do it.

File: src/gpuarray_buffer_cuda.c [hunk 9]
```diff
@@ -49,18 +49,72 @@ static int cuda_records(gpudata *, int, CUstream);
...
+    return NULL;
+  }
+  memcpy(k->bin_id, b->s, 64);
+  strb_appendn(k->src, b->s+64, b->l-64);
+  return k;
+}
+
+static int kernel_write(strb *res, strb *bin) {
+  strb_appendb(res, bin);
+  return strb_error(res);
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: What it should return? Normally a write tell the number of bytes written. I don't think this is what happen here.

File: src/gpuarray_buffer_cuda.c [hunk 10]
```diff
@@ -49,18 +49,72 @@ static int cuda_records(gpudata *, int, CUstream);
...
+}
+
+static strb *kernel_read(const strb *b) {
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I don't get what this do. Is the name correct? What it from from? It look more like a copy method then a read method.

File: src/gpuarray_buffer_cuda.c [hunk 11]
```diff
@@ -152,11 +208,38 @@ cuda_context *cuda_make_ctx(CUcontext ctx, int flags) {
...
+                                 (cache_eq_fn)strb_eq,
+                                 (cache_hash_fn)strb_hash,
                                  (cache_freek_fn)strb_free,
                                  (cache_freev_fn)cuda_freekernel);
   if (res->kernel_cache == NULL)
     goto fail_cache;
+
+  cache_path = getenv("GPUARRAY_CACHE_PATH");
+  if (cache_path != NULL) {
+    mem_cache = cache_lru(64, 8,
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: Can you explain this mem cache? What it is for?

File: src/private_cuda.h [hunk 12]
```diff
@@ -68,6 +68,7 @@ typedef struct _cuda_context {
   CUstream mem_s;
   gpudata *freeblocks;
   cache *kernel_cache;
+  cache *disk_cache;
```
Comment (nouiz) [inline src/private_cuda.h]: Why in the context? Do you have a sub path per arch?

File: src/gpuarray/error.h [hunk 13]
```diff
@@ -36,6 +36,7 @@ enum ga_error {
   GA_COMM_ERROR,
   GA_XLARGE_ERROR,
   GA_LOAD_ERROR,
+  GA_DEPRECATED_ERROR,
```
Comment (nouiz) [inline src/gpuarray/error.h]: This will request a version bump somewhere.

File: src/gpuarray_buffer_cuda.c [hunk 14]
```diff
@@ -996,22 +996,22 @@ static int detect_arch(const char *prefix, char *ret, CUresult *err) {
   return GA_NO_ERROR;
 }
 
-static void *call_compiler(const char *src, size_t len, const char *arch_arg,
-                           size_t *bin_len, char **log, size_t *log_len,
-                           int *ret) {
+static int call_compiler(cuda_context *ctx, strb *src, strb *ptx, strb *log) {
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: version bump.

File: src/gpuarray_buffer_cuda.c [hunk 15]
```diff
@@ -1020,41 +1020,115 @@ static void *call_compiler(const char *src, size_t len, const char *arch_arg,
...
 
-  buf = malloc(buflen);
-  if (buf == NULL) {
-    nvrtcDestroyProgram(&prog);
-    FAIL(NULL, GA_MEMORY_ERROR);
+  if (strb_ensure(ptx, buflen) == 0)
+    err = nvrtcGetPTX(prog, ptx->s+ptx->l);
+
+end:
+  nvrtcDestroyProgram(&prog);
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: Now you always destroy the program. Why? It the past it was only in case of error.

File: src/gpuarray_buffer_cuda.c [hunk 16]
```diff
@@ -1789,7 +1783,6 @@ const gpuarray_buffer_ops cuda_ops = {cuda_get_platform_count,
                                       cuda_freekernel,
                                       cuda_kernelsetarg,
                                       cuda_callkernel,
-                                      cuda_kernelbin,
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I think this break the ABI.

File: src/gpuarray/buffer.h (hunk 2 above)
Comment (abergeron) [inline src/gpuarray/buffer.h]: This might warrant a minor bump because there are new functions exposed in the API, yes.

File: src/gpuarray/config.h (hunk 3 above)
Comment (abergeron) [inline src/gpuarray/config.h]: It's mostly a cosmetic change since we use -fvisibility=hidden anyway.

File: bin/gpuarray-cache (hunk 5 above)
Comment (abergeron) [inline bin/gpuarray-cache]: It should work well on mac too, it's windows I have more doubts about.

Also, maybe there should be a default, I just didn't put one in the C code.

File: bin/gpuarray-cache (hunk 7 above)
Comment (abergeron) [inline bin/gpuarray-cache]: I want to be really sure :)

File: bin/gpuarray-cache (hunk 6 above)
Comment (abergeron) [inline bin/gpuarray-cache]: ok

File: bin/gpuarray-cache (hunk 4 above)
Comment (abergeron) [inline bin/gpuarray-cache]: ok

Comment (abergeron) [inline bin/gpuarray-cache]: Although we could do number of elements too.  Or age.  Or whatever.

File: src/gpuarray_buffer_cuda.c (hunk 9 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: This is not really a write.  It just appends to the strb buffer given.  So it returns if it worked or not.

File: src/gpuarray_buffer_cuda.c (hunk 14 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: This is an internal function so it has no impact on the API/ABI

File: src/gpuarray_buffer_cuda.c (hunk 10 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: In this case it does more or less a copy.  The point is that it has to fit into an API that asks it to read data from the input strb and deserialize the data.  It just happens that the data in this case is another strb.

File: src/gpuarray_buffer_cuda.c (hunk 16 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: No I moved that list to stritcly internal to avoid ABI breaks when it changes.

File: src/private_cuda.h (hunk 12 above)
Comment (abergeron) [inline src/private_cuda.h]: No, but I don't want to introduce contention on cache access and global locks.

File: src/gpuarray/error.h (hunk 13 above)
Comment (abergeron) [inline src/gpuarray/error.h]: Yes, an API bump probably.

File: src/gpuarray_buffer_cuda.c (hunk 11 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: It's to have an in-memory read cache for the disk cache.

**Comment by abergeron** [comment]
I finally got the Windows code working!

My test case of importing Theano (which compiles 6 blas kernels) goes from 5 to 2 seconds total time with this on Windows. On linux we go from 10 to 4 seconds (not the same machine).

There are still some details to tweak such as possibly setting a default value for the cache path and some code cleanup.

File: src/gpuarray_buffer_cuda.c [hunk 17]
```diff
@@ -937,41 +1021,123 @@ static void *call_compiler(const char *src, size_t len, const char *arch_arg,
...
-    buf = malloc(buflen);
-    if (buf == NULL) goto end2;
-    err2 = nvrtcGetProgramLog(prog, (char *)buf);
-    if (err2 != NVRTC_SUCCESS) goto end2;
-    if (log_len != NULL) *log_len = buflen;
-    *log = (char *)buf;
-    buf = NULL;
+  if (nvrtcGetProgramLogSize(prog, &buflen) == NVRTC_SUCCESS) {
+    strb_appends(log, "NVRTC compile log::\n");
+    if (strb_ensure(log, buflen) == 0)
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: This isn't enough. It should be buflen + 1 for line 1029

File: src/gpuarray_buffer_cuda.c [hunk 18]
```diff
@@ -937,41 +1021,123 @@ static void *call_compiler(const char *src, size_t len, const char *arch_arg,
...
   err = nvrtcGetPTXSize(prog, &buflen);
   if (err != NVRTC_SUCCESS) goto end;
 
-  buf = malloc(buflen);
-  if (buf == NULL) {
-    nvrtcDestroyProgram(&prog);
-    FAIL(NULL, GA_MEMORY_ERROR);
+  if (strb_ensure(ptx, buflen) == 0) {
+    err = nvrtcGetPTX(prog, ptx->s+ptx->l);
+    if (err == NVRTC_SUCCESS) ptx->l = buflen;
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I think it should be "ptx->l += buflen;". But in practice I think ptx->l will be zero most of the time. Still better safe then this hard to debug problem.

File: src/gpuarray_buffer_cuda.c [hunk 19]
```diff
@@ -937,41 +1021,123 @@ static void *call_compiler(const char *src, size_t len, const char *arch_arg,
...
+  err = make_bin(ctx, &ptx, bin, log);
+  if (err != GA_NO_ERROR) return err;
+  if (ctx->disk_cache) {
+    pk = memdup(&k, sizeof(k));
+    if (pk == NULL)
+      return GA_NO_ERROR;
+    cbin = strb_alloc(bin->l);
+    if (cbin == NULL) {
+      free(pk);
+      return GA_NO_ERROR;
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I don't like that. The cache is just skipped and the user don't have any information about that.

File: src/gpuarray_buffer_cuda.c [hunk 20]
```diff
@@ -937,41 +1021,123 @@ static void *call_compiler(const char *src, size_t len, const char *arch_arg,
...
+    }
+    strb_appendb(cbin, bin);
+    if (strb_error(cbin)) {
+      free(pk);
+      strb_free(cbin);
+      return GA_NO_ERROR;
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: idem

File: src/gpuarray_buffer_cuda.c [hunk 21]
```diff
@@ -1210,16 +1306,16 @@ static gpukernel *cuda_newkernel(gpucontext *c, unsigned int count,
...
     TAG_KER(res);
-    psb = memdup(&sb, sizeof(strb));
-    if (psb == NULL) {
-      cuda_freekernel(res);
-      strb_clear(&sb);
-      FAIL(NULL, GA_MEMORY_ERROR);
+    psrc = memdup(&src, sizeof(strb));
+    if (psrc != NULL) {
+      /* One of the refs is for the cache */
+      res->refcnt++;
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: Why it isn't the cache that add the ref if it succed?

File: src/private_cuda.h (hunk 12 above)
Comment (nouiz) [inline src/private_cuda.h]: Add that reason as a comment.

File: src/util/strb.h [hunk 22]
```diff
@@ -159,7 +159,21 @@ static inline void strb_appendb(strb *sb, strb *sb2) {
  *
  * A format error will place the strb in error mode.
  */
-GPUARRAY_LOCAL void strb_appendf(strb *, const char *f, ...);
+void strb_appendf(strb *, const char *f, ...);
+
+/*
+ * Reads from the file specified by the given file descriptor.
```
Comment (nouiz) [inline src/util/strb.h]: add that it read up to sz bytes.

File: src/util/strb.h [hunk 23]
```diff
@@ -159,7 +159,21 @@ static inline void strb_appendb(strb *sb, strb *sb2) {
...
+ *
+ * A read error will place the strb in error mode.
+ */
+void strb_read(strb *, int fd, size_t sz);
```
Comment (nouiz) [inline src/util/strb.h]: I think there is an error in this line that prevent compilation... why only "strb *," and no name for that parameter? Why does jenkins passed?

File: bin/gpuarray-cache [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+    for root, dirs, files in os.walk(os.environ.get('GPUARRAY_CACHE',
+                                                    '~/.gpuarray/cache/')):
+        for file in files:
+            fpath = os.path.join(root, file)
+            st = os.stat(fpath)
+            content.append((st.st_atime, st.st_size, fpath))
+
+    content.sort()
+    cur_size = 0
+    for _, size, path in content:
```
Comment (abergeron) [inline bin/gpuarray-cache]: content.sort() does that since I store the tuples with (atime, size, path)

File: src/gpuarray_blas_opencl_clblas.c [hunk 25]
```diff
@@ -100,7 +100,7 @@ static int sgemmBatch(cb_order order, cb_transpose transA, cb_transpose transB,
     ARRAY_INIT(C[i]);
     err = clblasSgemm(convO(order), convT(transA), convT(transB), M, N, K,
                       alpha, A[i]->buf, offA[i], lda, B[i]->buf, offB[i], ldb,
-                      beta, C[i]->buf, offB[i], ldc, 1, &ctx->q,
+                      beta, C[i]->buf, offC[i], ldc, 1, &ctx->q,
```
Comment (abergeron) [inline src/gpuarray_blas_opencl_clblas.c]: Yes, it's a bugfix.

File: make.bat [hunk 26]
```diff
@@ -0,0 +1,6 @@
+del bld
```
Comment (abergeron) [inline make.bat]: ok

File: src/gpuarray/buffer.h (hunk 2 above)
Comment (abergeron) [inline src/gpuarray/buffer.h]: I'll do it.

File: bin/gpuarray-cache (hunk 8 above)
Comment (abergeron) [inline bin/gpuarray-cache]: I simplified this code.

File: src/gpuarray_buffer_cuda.c (hunk 17 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: We are cutting the final '\0' of the log at line 1028 so there will be enough space.

Also even if there wasn't enough space strb_appendc() will extend the buffer if required.

File: src/gpuarray_buffer_cuda.c (hunk 18 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: Yes.

File: src/gpuarray_buffer_cuda.c (hunk 19 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: Would you prefer to error out in case we can't add the kernel to the cache for some reason?

I can make that change.

File: src/gpuarray_buffer_cuda.c (hunk 21 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: Because I would have to teach the cache how to add references (which is not possible for just any object) where as now I just have to teach it how to free objects.

File: src/private_cuda.h (hunk 12 above)
Comment (abergeron) [inline src/private_cuda.h]: ok

File: src/util/strb.h (hunk 22 above)
Comment (abergeron) [inline src/util/strb.h]: ok

File: src/util/strb.h (hunk 23 above)
Comment (abergeron) [inline src/util/strb.h]: We don't need to specify parameter names on declaration, so this wouldn't be a problem.  However it jars with the style of the rest of the file so I'll fix it.

File: src/gpuarray_buffer_cuda.c (hunk 17 above)
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: Add on line 1028, that this is what that line does

File: src/gpuarray_buffer_cuda.c (hunk 19 above)
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: I would prefer a warning, but I think it isn't easy. So a memory error seem more appropriate as this is the only cause of this that I can think of.

File: src/gpuarray_buffer_cuda.c (hunk 21 above)
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: ok

File: src/cache/disk.c [hunk 27]
```diff
@@ -0,0 +1,459 @@
...
+  cache * mem;
+  kwrite_fn kwrite;
+  vwrite_fn vwrite;
+  kread_fn kread;
+  vread_fn vread;
+  const char *dirp;
+} disk_cache;
+
+
+static unsigned long long ntohull(const char *_in) {
```
Comment (abergeron) [inline src/cache/disk.c]: Like ntohl but for unsigned long long (64 bits).

It converts from network byte order to host byte order.

File: src/cache/disk.c [hunk 28]
```diff
@@ -0,0 +1,459 @@
...
+  out[1] = (unsigned char)(in >> 48);
+  out[2] = (unsigned char)(in >> 40);
+  out[3] = (unsigned char)(in >> 32);
+  out[4] = (unsigned char)(in >> 24);
+  out[5] = (unsigned char)(in >> 16);
+  out[6] = (unsigned char)(in >> 8);
+  out[7] = (unsigned char)(in);
+}
+
+static int catp(char *path, const char *dirp, const char *rpath) {
```
Comment (abergeron) [inline src/cache/disk.c]: ok

File: src/cache/disk.c [hunk 29]
```diff
@@ -0,0 +1,459 @@
...
+    return -1;
+  }
+  if (strlcat(path, rpath, PATH_MAX) >= PATH_MAX) {
+    errno = ENAMETOOLONG;
+    return -1;
+  }
+  return 0;
+}
+
+static int openp(const char *dirp, const char *rpath, int flags, int mode) {
```
Comment (abergeron) [inline src/cache/disk.c]: Ok.

File: src/gpuarray_buffer_cuda.c (hunk 17 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: ok

File: src/gpuarray_buffer_cuda.c (hunk 19 above)
Comment (abergeron) [inline src/gpuarray_buffer_cuda.c]: With disk_cache it could also be a write error of some kind.

**Comment by nouiz** [comment]
I have a segfault when I try this PR with the cache enabled, but not when
the cache isn't enabled. This is for the pygpu tests, the cache was reusing
the output of libgpuarray tests.

**Comment by nouiz** [comment]
I also have the segfaul of pygpu tests with an empty cache. In that case
the segfaul happen earlier.

I let you investigate this more.

File: src/cache/disk.c (hunk 27 above)
Comment (nouiz) [inline src/cache/disk.c]: The name is very cryptic. If you check htonull, this made me think we talk about null ptr!

Can you just add for both a 1 line comment that tell what you wrote here.

**Comment by abergeron** [comment]
I had to rebase for some unrelated problems, but I found and fixed the cause of the segfault.

The whole test suite at once passes.  Just for fun I made a second run of the testsuite with the cache full and it took 52s instead of 700s.

**Comment by abergeron** [comment]
What is remaining:
 - Decide if erroring out when there are problems adding to the cache is a good idea
 - Maybe add some versioning to the caching for debug/non-debug, new cuda releases.

**Comment by abergeron** [comment]
Also, in release mode the tests go from 542 seconds to 30.

**Comment by nouiz** [comment]
How does it handle NFS? I ran the libgpuarray and pygpu tests in different config and they pass. So the only thing left for me is to understand how NFS is handled

File: src/gpuarray_buffer_cuda.c [hunk 30]
```diff
@@ -50,10 +50,17 @@ static int detect_arch(const char *prefix, char *ret, CUresult *err);
...
+  uint8_t version;
+  uint8_t debug;
+  uint8_t major;
+  uint8_t minor;
+  uint32_t reserved;
   char bin_id[64];
   strb src;
 } kernel_key;
 
+#define KERNEL_KEY_MM (sizeof(kernel_key) - sizeof(strb))
```
Comment (nouiz) [inline src/gpuarray_buffer_cuda.c]: What MM mean? Add this in comment.

**Comment by nouiz** [comment]
With the cache, test_elemwise fail:
```
$ GPUARRAY_CACHE_PATH=~/.theano/gpuarray_cache/ DEVICE=cuda0 make test
Running tests...
Test project /u/bastienf/repos/libgpuarray/Build
      Start  1: test_types
 1/11 Test  #1: test_types .......................   Passed    0.01 sec
      Start  2: test_util
 2/11 Test  #2: test_util ........................   Passed    0.01 sec
      Start  3: test_util_integerfactoring
 3/11 Test  #3: test_util_integerfactoring .......   Passed    0.51 sec
      Start  4: test_reduction
 4/11 Test  #4: test_reduction ...................   Passed    5.91 sec
      Start  5: test_array
 5/11 Test  #5: test_array .......................   Passed    2.37 sec
      Start  6: test_blas
 6/11 Test  #6: test_blas ........................   Passed    4.28 sec
      Start  7: test_elemwise
 7/11 Test  #7: test_elemwise ....................***Failed    9.82 sec
      Start  8: test_error
 8/11 Test  #8: test_error .......................   Passed    0.01 sec
      Start  9: test_buffer
 9/11 Test  #9: test_buffer ......................   Passed    2.46 sec
      Start 10: test_buffer_collectives
10/11 Test #10: test_buffer_collectives ..........   Passed    0.83 sec
      Start 11: test_collectives
11/11 Test #11: test_collectives .................   Passed    0.55 sec

91% tests passed, 1 tests failed out of 11

Total Test time (real) =  26.84 sec

The following tests FAILED:
          7 - test_elemwise (Failed)
Errors while running CTest
Makefile:84: recipe for target 'test' failed
make: *** [test] Error 8
```

**Comment by abergeron** [comment]
I fixed the bug that made the test fail.

**Comment by nouiz** [comment]
I,ll try again tomorrow, but I didn't saw a speed up in pygpu tests with the current state of the PR with this line:

GPUARRAY_DISK_CACHE=~/.theano/gpuarray_cache DEVICE=cuda0 python -c "import pygpu;pygpu.test()"

**Comment by abergeron** [comment]
Yes, because the name of the variable is GPUARRAY_CACHE_PATH, not GPUARRAY_DISK_CACHE
